R/fct_cross.R

```diff
@@ -33,6 +33,13 @@ fct_cross <- function(.f, ..., sep = ":", keep_empty = FALSE) {
       rlang::invoke(expand.grid, all_levels),
       sep = sep
     ))
+  } else if (preserve_orders) {
+    all_levels <- lapply(.data, levels)
```

Comment:
I think the way to fix this is to generate the full set of levels outside the if the statement, and the filter them to the set of actual levels (using `intersect()`, which is a shortcut for your `%in%` pattern) if `keep_empty` is `FALSE`.